### DISCUSSION TITLE
Re-scan all cookies against whitelist?

### DISCUSSION THREAD

**Discussion by gbhug5a** [Cookie-AutoDelete/Cookie-AutoDelete#1106] [Q&A / Support] [answered]
If CAD has been disabled for a while to get a site to run properly, then is re-enabled, I expect CAD to remove all the cookies which were added during the disabled period which are not on the whitelist But it does not appear to do that. Is there a way to get CAD to run ALL existing cookies against the whitelist, and delete those that aren't whitelisted? If not on every browser startup, could it at least do that when it is switched from disabled to enabled.

If it is supposed to be doing that already, what setting might I have set wrong? I have Automatic Cleanup turned on.

Edit: Firefox under Windows 10 20H2, if it matters.

**Comment by kennethtran93** [comment]
Automatic cleaning is only triggered after a tab close, domain change and/or browser startup, depending on the configuration. If you want to trigger a cleanup you can do so via the CAD popup by clicking the Yellow 'Clean' button. Note that domains in opened tabs are not cleaned up by default.

**Reply by gbhug5a** [reply]
Thanks very much for the reply.  Well, it definitely hasn't been working on a browser startup.  I manually deleted several dozen months-old cookies this morining.  I don't see a configuration option for browser startup other than one related to the Greylist, and I only have a Whitelist.  But let me make sure I understand what I should do:

If I open Firefox to a blank page, then go into options and re-enable CAD, then click on the yellow Clean button, and select the first option (Clean, including open tabs), it will clean all the cookies from all sites that don't match the whitelist.  Is that correct?

**Reply by kennethtran93** [reply]
The clean button itself will perform regular cleanup, excluding open tabs. The dropdown contains the additional options, including cleaning the opened tabs as well. 

Note that CAD automatic cleanup must be enabled before browser exit/restart in order for browser cleanup on startup to take effect. 

Which Firefox type are you using?
What is your browser privacy - enhanced tracking protection setting at?

**Reply by gbhug5a** [reply]
I've had automatic cleanup enabled for years.  And I close the browser several times a day.  Having spent the better part of half an hour this morning going through all my cookies and manually deleting the dozens that shouldn't be there, some dating back months, it's clear to me that the universal Clean function is not running on browser startup, and hasn't been running in some time.  I'm sure those cookies were placed during periods when I had CAD disabled, but they should have been cleared out on the first browser startup after CAD had been re-enabled.  But it doesn't really matter so long as the manual Clean will get rid of them.  Of course I'll have to test that at some point.

I'm running Firefox x64 for Windows desktop, currently at v87.0.  I've been running the Strict option for advanced tracking protection, but am concluding that I probably need to set it back to Standard.  Too many sites don't work when it's on Strict, even if I set them individually back to Standard.

**Reply by kennethtran93** [reply]
Unfortunately the strict setting in Firefox 86+ is breaking webextension like this one. It is an issue on the Firefox side. This is being tracked through #807.

**Comment by kennethtran93** [comment, marked as answer]
Due to the user's configuration of using strict mode in Firefox 86+, this is now a duplicate issue of #807.

**Reply by gbhug5a** [reply]
There may be something else going on.  I've set the mode back to Standard, then rebooted, then did a manual Clean.  But after all that I still have a cookie from alibaba.com even though there's no such entry on the Expressions list.  I missed that one this morning, and it shows as being 17 days old.  I don't understand why that cookie is still there.

**Reply by kennethtran93** [reply]
If you feel adventurous enough, there is the debug mode option available, but you'll need to scan through a lot of debugging entries.

Is it a cookie or remaining site data?  these are two separate items.

**Reply by gbhug5a** [reply]
Firefox shows the albaba.com entry as two cookies, with no site data.  It shows the cookies as 18 days old.  But I don't think there's a way to look at cookie contents in Firefox, so I don't know what's inside.

I wouldn't know what to do with the debug mode.

![Cookie](https://user-images.githubusercontent.com/30031827/113224234-99c81e00-9250-11eb-8ea9-1b9b5b5aa253.jpg)

**Reply by kennethtran93** [reply]
There are webextensions designed to look into each cookie details, though deleting them from there may face the same issue that were facing.

**Reply by gbhug5a** [reply]
It's not clear from the Issue 807 discussion if there's a way to prevent this problem while waiting for Mozilla to fix it.  Is it enough to go back Standard mode?  Are there other about:config settings that need to be set a certain way?

**Reply by kennethtran93** [reply]
In order to prevent this problem from happening, you would need to disable Firefox's newest Cookie Protection introduced in 86, which for the most part going to standard may be sufficient, though existing cookies data generated by Firefox will most likely stay the way it is so I cannot confirm if we would be able to remove those cookies after configuration changes, mainly due to API limitations.

I don't have a clear answer about preventing the problem from happening other than to wait for Firefox to patch it, or to downgrade your version of firefox.  You cannot, at this time, have a working WebExtension with Firefox's new Cross-Site Cookie Isolation/Protection.

The Cross-site cookie isolations appear to simply set `network.cookie.cookieBehavior` to **5**, whereas pre-86 in strict mode I believe it was set to **4** to block/isolate cross-site tracking cookies.

You may also use the 'custom' privacy setting, and set the cookie dropdown value to use cross site tracking instead of just cross site and that should do the same thing as modifying the cookieBehavior to **4**, keeping the rest checked.
![image](https://user-images.githubusercontent.com/6724477/113372244-10cae880-931d-11eb-8864-8daf2a162d41.png)

I'm not sure if enabling `privacy.firstparty.isolate` will do much, but I believe I read somewhere (cannot recall source at this time) that this would override the usage of dynamic first party isolation as well.  Could have been mistaken it for something else though.

**Reply by gbhug5a** [reply]
I've gone back to Standard, and that sets network.cookie.cookieBehavior to 4, and privacy.firstparty.isolate to false.  And the alibaba cookie is still there.  So it appears I would need to delete all cookies in the FF UI, and start over.  Currently, the log continues to say it removed 72 cookies each time I boot the browser, so I assume those are more tainted cookies that I missed when doing manual deletes, and which CAD is not able to delete because of this issue.   I've had several problems with Strict that weren't resolved by setting the individual site back to Standard, so I think I will stay with Standard anyway.  But I would rather not delete all cookies unless I'm sure that would mitigate the problem.

However, I'm not sure this just relates to the new Firefox isolation policy.  Some of the cookies I deleted manually dated back two months or more.

**Reply by kennethtran93** [reply]
It would certainly take more debug work, though would be harder as that would require me to attempt to follow in your footsteps.  There has been one where a cookie has been created with a question mark in its path to which both Chrome and Firefox handles the creation differently.  While it was created in Firefox, it could not be removed through the webextension API, and I did post a bugzilla report a while back but nothing has been done to date about it.

Due to the way CAD is coded right now, we do not check whether or not the cookie has been deleted as we use Promise.all and pass in the cookies to be removed.  We presume it has been removed but as you see that is not always the case.  I may switch to Promise.allSettled to allow for checks in the future.


Note that we can only use available APIs that Firefox has given us.  The internal settings and functions are not available to us, as clearly noted with Firefox cookie storage being able to remove what we can't.
